[RELENG-1011] - Add AuthenticodeEV signing vars

ahal opened the pull request.

ahal committed:
[RELENG-1011] - Add Authenticode EV signing vars and certificates

JohanLorenzo commented on signingscript/src/signingscript/sign.py:
```diff
@@ -1405,10 +1405,10 @@ async def signer(digest, digest_algo):
-    if "authenticode_sha2" in fmt:
-        digest_algo = "sha256"
-    else:
+    if fmt in ["autograph_authenticode", "autograph_authenticode_stub"]:
```
Looks like we don't use these formats anymore on central[1]:

<img width="583" alt="image" src="https://user-images.githubusercontent.com/5907366/212023904-f9ba1aa7-27e6-4064-8b75-2ff51395ec79.png">

Same results with ESR 102 and 91. ESR 78 was still referencing them[2], though:

<img width="684" alt="image" src="https://user-images.githubusercontent.com/5907366/212024531-54cfc68a-4b93-4129-a308-9c37015dc226.png">

I wonder if it's worth retiring these signing format in a followup PR.

[1] https://searchfox.org/mozilla-central/search?path=&q=autograph_authenticode
[2] https://searchfox.org/mozilla-esr78/search?path=&q=autograph_authenticode

ahal replied:
+1 to simplification! This stuff is already too complex for me. Though I'm definitely not going to tackle it in this PR.

JohanLorenzo commented on signingscript/docker.d/passwords.yml:
```diff
@@ -413,6 +413,11 @@ in:
+          - ["https://autograph-external.prod.autograph.services.mozaws.net",
+             {"$eval": "AUTOGRAPH_AUTHENTICODE_EV_USERNAME"},
+             {"$eval": "AUTOGRAPH_AUTHENTICODE_EV_PASSWORD"},
+             ["autograph_authenticode_file"]
```
Nit (no need to fix this in this current PR): I wonder if `autograph_authenticode_file` can confuse future readers knowing that `autograph_authenticode_sha2` and `autograph_authenticode_sha2_stub` also exist. That said, I know this change requires coordination so no need to address it right away.

ahal replied:
Maybe it's worth creating a single entrypoint that can handle both single files as well as zip files...

JohanLorenzo reviewed (approved):
Looks good to me! 👍

jcristau commented on signingscript/src/signingscript/task.py:
```diff
@@ -53,6 +54,7 @@
         "autograph_authenticode_sha2": sign_authenticode_zip,
         "autograph_authenticode_sha2_stub": sign_authenticode_zip,
         "autograph_authenticode_sha2_rfc3161_stub": sign_authenticode_zip,
+        "autograph_authenticode_file": sign_authenticode_file,
```
I doubt we can re-use sign_authenticode_file as-is.  It is currently set to use `context.config["authenticode_ca"]`, which is not the same intermediate that the EV cert chains to, and one of the `authenticode_cert*` config options as signing cert, which also point at the non-EV cert.
I'm seeing 2 options from a quick look:
- a new sign_authenticode_ev function that passes in the right EE and intermediate certs to winsign directly
- a new "keyid" for sign_authenticode_file, plus making it possible to pass in a different intermediate cert in e.g. `authenticode_ca_{keyid}` instead of always `authenticode_ca`
We'll also want to add both certs to signingscript/src/signingscript/data.

ahal replied:
Thanks, appreciate the pointer! I was pretty sure this was going to fail, but wasn't sure what I needed to do. This will help save me a lot of time investigating :)

jcristau reviewed (changes requested):
Requesting changes because I *think* this isn't quite right.

ahal committed:
Rename 'sign_autograph_zip' to make it clear it also supports single files

This funcion supports both single files and zip files, but the name and
docstring imply that only the latter is supported. This simple rename
should help make it more clear.

ahal committed:
Switch authenticode signing to newer sha256 digest by default

We shouldn't be using the sha1 algorithm for any new signing formats. So
rather than looking for sha2 in the name, let's special case the only
two formats that still need to use sha1, and default everything else to
sha256.

ahal committed:
Allow 'keyid's in intermediate authenticode certs

The upcoming Authenticode EV cert uses a different intermediate cert
from the other authenticode formats. When using a keyid,
`sign_authenticode` will now check if a ca cert with that keyid is
configured first. If none are found, it will fallback to the original
one.

This way we can support keyids in intermediate certs while preserving
backwards compatibility.

ahal commented on signingscript/docker.d/passwords.yml:
```diff
@@ -413,6 +413,12 @@ in:
+             {"$eval": "AUTOGRAPH_AUTHENTICODE_EV_USERNAME"},
+             {"$eval": "AUTOGRAPH_AUTHENTICODE_EV_PASSWORD"},
+             ["autograph_authenticode"],
+             "ev"
```
I see the "202005" keyid is used above, but I'm a bit confused about how this works..

From what I can tell, these keyids should end up getting parsed into [this property](https://github.com/mozilla-releng/scriptworker-scripts/blob/ffe4fd9391d8857d17b4840d32acd9d98aa0d994/signingscript/src/signingscript/utils.py#L25). But from the looks of it, that property is only ever accessed in [this one place](https://github.com/mozilla-releng/scriptworker-scripts/blob/ffe4fd9391d8857d17b4840d32acd9d98aa0d994/signingscript/src/signingscript/sign.py#L1020).

But `sign_authenticode_file` doesn't ever call into that function and instead extracts keyids out of the format [over here](https://github.com/mozilla-releng/scriptworker-scripts/blob/ffe4fd9391d8857d17b4840d32acd9d98aa0d994/signingscript/src/signingscript/sign.py#L1397). And I can't seem to find anywhere that is joining a `keyid` to the format!

So how do authenticode keyids work??

ahal replied:
I traced the signing format backwards, and afaict, it looks like it might just need to come from the [task payload](https://github.com/mozilla-releng/scriptworker-scripts/blob/ffe4fd9391d8857d17b4840d32acd9d98aa0d994/signingscript/src/signingscript/task.py#L224)? However, if that's the case, what's the point of putting the signing format in `passwords.yml` (like we do with the "202005" one)? Maybe that's actually a no-op and when :catlee landed #218 he didn't actually switch anything over... Fwiw there are no references to "202005" in Gecko.

ahal replied:
Sigh nevermind. My initial statement was faulty and `sign_authenticode_file` does indeed ultimately call into `sign_with_autograph`.

On the bright side, I've learned a lot about how this works. And on the even brighter side, the PR looks like it might actually work.

jcristau reviewed (approved):
This looks good to me, thanks!.  Not a guarantee it'll work but worth giving it a try.
